File: src/teleglance/media.py
"""Async media downloading."""

from __future__ import annotations

import mimetypes
import os
import tempfile
from pathlib import Path
from urllib.parse import urlparse

from .errors import DownloadError
from .models.media import Media
from .transport import TransportProtocol


def resolve_url(media: Media | str) -> str:
    """Best downloadable URL for a media object, or raise DownloadError."""
    if isinstance(media, str):
        return media
    url = getattr(media, "url", None)
    if isinstance(url, str) and url:
        return url
    kind = getattr(media, "type", type(media).__name__)
    if kind == "document":
        raise DownloadError("t.me previews do not expose direct URLs for document attachments")
    raise DownloadError(f"{kind} media has no downloadable URL")


def _filename_for(url: str, content_type: str | None) -> str:
    name = Path(urlparse(url).path).name
    if name and "." in name:
        return name
    ext = mimetypes.guess_extension((content_type or "").split(";")[0].strip()) or ".bin"
    return (name or "media") + ext


def _validate_max_bytes(max_bytes: int | None) -> None:
    if max_bytes is not None and max_bytes <= 0:
        raise ValueError("max_bytes must be positive")


def _declared_too_large(content_length: str | None, max_bytes: int | None) -> bool:
    if content_length is None or max_bytes is None:
        return False
    try:
        return int(content_length) > max_bytes
    except ValueError:
        return False


async def download_bytes(
    transport: TransportProtocol,
    media: Media | str,
    *,
    max_bytes: int | None = None,
) -> bytes:
    _validate_max_bytes(max_bytes)
    url = resolve_url(media)
    async with transport.stream(url) as response:
        if response.status_code != 200:
            raise DownloadError(f"GET {url} returned {response.status_code}")
        if _declared_too_large(response.headers.get("Content-Length"), max_bytes):
            raise DownloadError(f"download exceeds max_bytes={max_bytes}")
        chunks = []
        size = 0
        async for chunk in response.aiter_bytes():
            size += len(chunk)
            if max_bytes is not None and size > max_bytes:
                raise DownloadError(f"download exceeds max_bytes={max_bytes}")
            chunks.append(chunk)
        return b"".join(chunks)


async def download_media(
    transport: TransportProtocol,
    media: Media | str,
    dest: str | Path | None = None,
    *,
    filename: str | None = None,
    overwrite: bool = False,
    max_bytes: int | None = None,
) -> Path:
    """Download to disk. ``dest`` may be a directory (filename derived from the
    URL / content type) or a full file path. Defaults to the current directory."""
    _validate_max_bytes(max_bytes)
    url = resolve_url(media)
    async with transport.stream(url) as response:
        if response.status_code != 200:
            raise DownloadError(f"GET {url} returned {response.status_code}")

        if _declared_too_large(response.headers.get("Content-Length"), max_bytes):
            raise DownloadError(f"download exceeds max_bytes={max_bytes}")

        dest_path = Path(dest) if dest is not None else Path.cwd()
        if dest_path.suffix and not dest_path.is_dir():
            target = dest_path
        else:
            dest_path.mkdir(parents=True, exist_ok=True)
            content_type = response.headers.get("Content-Type")
            safe_filename = Path(filename).name if filename else _filename_for(url, content_type)
            target = dest_path / safe_filename

        target.parent.mkdir(parents=True, exist_ok=True)
        if target.exists() and not overwrite:
            raise DownloadError(f"destination already exists: {target}")
        temporary_path: Path | None = None
        try:
            with tempfile.NamedTemporaryFile(
                mode="wb", dir=target.parent, prefix=f".{target.name}.", delete=False
            ) as fh:
                temporary_path = Path(fh.name)
                size = 0
                async for chunk in response.aiter_bytes():
                    size += len(chunk)
                    if max_bytes is not None and size > max_bytes:
                        raise DownloadError(f"download exceeds max_bytes={max_bytes}")
                    fh.write(chunk)
                fh.flush()
                os.fsync(fh.fileno())
            if target.exists() and not overwrite:
                raise DownloadError(f"destination already exists: {target}")
            temporary_path.replace(target)
        finally:
            if temporary_path is not None:
                temporary_path.unlink(missing_ok=True)
    return target

File: packages/ts/src/media.ts
import { promises as fs } from 'fs';
import { dirname, basename, extname } from 'path';
import { randomBytes } from 'crypto';
import type { TransportProtocol } from './transport.js';
import { DownloadError } from './errors.js';
import type { Media } from './models/media.js';

export function resolveUrl(media: Media | string): string {
  if (typeof media === 'string') return media;
  const url = (media as any).url;
  if (typeof url === 'string' && url) return url;
  const kind = media.type;
  if (kind === 'document') {
    throw new DownloadError('t.me previews do not expose direct URLs for document attachments');
  }
  throw new DownloadError(`${kind} media has no downloadable URL`);
}

function filenameFor(url: string, contentType: string | null): string {
  const urlPath = new URL(url, 'https://t.me').pathname;
  let pathName = basename(urlPath);
  let name: string;

  if (pathName && pathName.includes('.')) {
    name = pathName;
  } else {
    const mime = contentType?.split(';')[0]?.trim();
    const ext = mime ? extFromMime(mime) : '.bin';
    name = (pathName || 'media') + ext;
  }

  if (name.length > 200) {
    const ext = extname(name);
    const stemLimit = 200 - ext.length;
    const stem = basename(name, ext);
    name = stem.slice(0, stemLimit) + ext;
  }

  return name;
}

function extFromMime(mime: string): string {
  const map: Record<string, string> = {
    'image/jpeg': '.jpg',
    'image/png': '.png',
    'image/gif': '.gif',
    'image/webp': '.webp',
    'video/mp4': '.mp4',
    'audio/ogg': '.ogg',
    'audio/mpeg': '.mp3',
  };
  return map[mime] || '.bin';
}

function validateMaxBytes(maxBytes: number | undefined): void {
  if (maxBytes !== undefined && maxBytes <= 0) {
    throw new Error('maxBytes must be positive');
  }
}

function declaredTooLarge(contentLength: string | undefined, maxBytes: number | undefined): boolean {
  if (!contentLength || maxBytes === undefined) return false;
  const size = parseInt(contentLength, 10);
  return !isNaN(size) && size > maxBytes;
}

export async function downloadBytes(
  transport: TransportProtocol,
  media: Media | string,
  maxBytes?: number,
): Promise<Buffer> {
  validateMaxBytes(maxBytes);
  const url = resolveUrl(media);
  const streamData = await transport.stream(url);
  const { statusCode, headers, body } = streamData;

  if (statusCode !== 200) {
    throw new DownloadError(`GET ${url} returned ${statusCode}`);
  }

  const contentLength = Array.isArray(headers['content-length'])
    ? headers['content-length'][0]
    : headers['content-length'];

  if (declaredTooLarge(contentLength, maxBytes)) {
    throw new DownloadError(`Download exceeds maxBytes=${maxBytes}`);
  }

  const chunks: Buffer[] = [];
  let size = 0;

  for await (const chunk of body) {
    const buf = Buffer.from(chunk);
    size += buf.length;
    if (maxBytes !== undefined && size > maxBytes) {
      throw new DownloadError(`Download exceeds maxBytes=${maxBytes}`);
    }
    chunks.push(buf);
  }

  return Buffer.concat(chunks);
}

export async function downloadMedia(
  transport: TransportProtocol,
  media: Media | string,
  dest?: string,
  filename?: string,
  overwrite?: boolean,
  maxBytes?: number,
): Promise<string> {
  validateMaxBytes(maxBytes);
  const url = resolveUrl(media);
  const streamData = await transport.stream(url);
  const { statusCode, headers, body } = streamData;

  if (statusCode !== 200) {
    throw new DownloadError(`GET ${url} returned ${statusCode}`);
  }

  const contentLength = Array.isArray(headers['content-length'])
    ? headers['content-length'][0]
    : headers['content-length'];

  if (declaredTooLarge(contentLength, maxBytes)) {
    throw new DownloadError(`Download exceeds maxBytes=${maxBytes}`);
  }

  const destPath = dest || process.cwd();
  let target: string;

  try {
    const stats = await fs.stat(destPath);
    if (stats.isDirectory()) {
      const contentType = Array.isArray(headers['content-type'])
        ? headers['content-type'][0]
        : headers['content-type'];
      const safeFilename = filename ? basename(filename) : filenameFor(url, contentType || null);
      target = `${destPath}/${safeFilename}`;
    } else {
      target = destPath;
    }
  } catch {
    if (destPath.includes('.') && !destPath.endsWith('/')) {
      target = destPath;
    } else {
      await fs.mkdir(destPath, { recursive: true });
      const contentType = Array.isArray(headers['content-type'])
        ? headers['content-type'][0]
        : headers['content-type'];
      const safeFilename = filename ? basename(filename) : filenameFor(url, contentType || null);
      target = `${destPath}/${safeFilename}`;
    }
  }

  await fs.mkdir(dirname(target), { recursive: true });

  if (!overwrite) {
    try {
      await fs.access(target);
      throw new DownloadError(`Destination already exists: ${target}`);
    } catch (err: any) {
      if (err instanceof DownloadError) throw err;
    }
  }

  const tempPath = `${dirname(target)}/.${basename(target)}.${randomBytes(8).toString('hex')}.tmp`;
  let size = 0;

  try {
    const writeStream = await fs.open(tempPath, 'w');
    try {
      for await (const chunk of body) {
        const buf = Buffer.from(chunk);
        size += buf.length;
        if (maxBytes !== undefined && size > maxBytes) {
          throw new DownloadError(`Download exceeds maxBytes=${maxBytes}`);
        }
        await writeStream.write(buf);
      }
      await writeStream.sync();
    } finally {
      await writeStream.close();
    }

    if (!overwrite) {
      try {
        await fs.access(target);
        throw new DownloadError(`Destination already exists: ${target}`);
      } catch (err: any) {
        if (err instanceof DownloadError) throw err;
      }
    }

    await fs.rename(tempPath, target);
  } catch (err) {
    await fs.unlink(tempPath).catch(() => {});
    throw err;
  }

  return target;
}
